pkg/steps/clusterinstall/clusterinstall.go

```diff
@@ -0,0 +1,140 @@
+package clusterinstall
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/ghodss/yaml"
+
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	coreclientset "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	templateapi "github.com/openshift/api/template/v1"
+
+	"github.com/openshift/ci-operator/pkg/api"
+	"github.com/openshift/ci-operator/pkg/junit"
+	"github.com/openshift/ci-operator/pkg/steps"
+)
+
+type e2eTestStep struct {
+	config     api.OpenshiftInstallerClusterTestConfiguration
+	testConfig api.TestStepConfiguration
+
+	secretClient coreclientset.SecretsGetter
+	jobSpec      *api.JobSpec
+
+	step api.Step
+	nestedSubTests
+}
+
+type nestedSubTests interface {
+	SubTests() []*junit.TestCase
+}
+
+// E2ETestStep installs a cluster and then runs end-to-end tests against it.
+func E2ETestStep(
+	config api.OpenshiftInstallerClusterTestConfiguration,
+	testConfig api.TestStepConfiguration,
+	params api.Parameters,
+	podClient steps.PodClient,
+	templateClient steps.TemplateClient,
+	secretClient coreclientset.SecretsGetter,
+	artifactDir string,
+	jobSpec *api.JobSpec,
+) api.Step {
+	var template *templateapi.Template
+	if err := yaml.Unmarshal([]byte(installTemplateE2E), &template); err != nil {
+		panic(fmt.Errorf("the embedded template is invalid: %v", err))
+	}
+
+	template.Name = testConfig.As
+
+	if config.Upgrade {
+		overrides := make(map[string]string)
+		for i := range template.Parameters {
+			p := &template.Parameters[i]
+			switch p.Name {
+			case "JOB_NAME_SAFE":
+				if _, ok := os.LookupEnv(p.Name); !ok {
+					overrides[p.Name] = testConfig.As
+				}
+			case "TEST_COMMAND":
+				p.Value = testConfig.Commands
+			case "CLUSTER_TYPE":
+				p.Value = strings.Split(string(config.ClusterProfile), "-")[0]
+			case "IMAGE_INSTALLER":
+				if _, ok := os.LookupEnv(p.Name); !ok {
+					overrides[p.Name] = "stable-initial:installer"
+				}
+			case "INSTALL_INITIAL_RELEASE":
+				if _, ok := os.LookupEnv(p.Name); !ok {
+					overrides[p.Name] = "true"
+				}
+			}
+		}
+		params = api.NewOverrideParameters(params, overrides)
+	}
+
+	step := steps.TemplateExecutionStep(template, params, podClient, templateClient, artifactDir, jobSpec)
+	subTests, ok := step.(nestedSubTests)
+	if !ok {
+		panic(fmt.Sprintf("unexpected %T", step))
+	}
+
+	return &e2eTestStep{
+		config:     config,
+		testConfig: testConfig,
+
+		secretClient: secretClient,
+		jobSpec:      jobSpec,
+
+		step:           step,
+		nestedSubTests: subTests,
+	}
+}
+
+func (s *e2eTestStep) checkPrereqs() error {
+	return nil
+}
+
+func (s *e2eTestStep) Inputs(ctx context.Context, dry bool) (api.InputDefinition, error) {
+	return nil, nil
```

Comment:
???

Reply:
The e2e test step doesn't load any unique input from the graph that would alter the hash used to guarantee unique namespaces.